standards/practices/testing.md

```diff
@@ -0,0 +1,198 @@
+        // Act
+        // Try to log in.
+        $response = $this->post('/login', [
+            'email'    => $user->email,
+            'password' => 'pw1234'
+        ]);
+
+        // Assert
+        // Ensure user logged in.
+        $response->assertRedirect(route('home'));
+        $this->assertAuthenticatedAs($user);
+    }
+}
+```
+
+*In most frameworks, it's not worth testing authentication, because the framework handles pretty much everything. However in this example, the authentication is customized enough that we need to avoid regressions.*
+
+These two tests improve on the previous example in several ways:
+
+1. We converted from a unit test to an integration test (called a feature test in Laravel). The feature depends on several systems more simply tested with a couple short integration tests than with several unit tests that can't reliably test the whole feature.
+2. Each test covers only one action.
+3. We formatted the test names so people can easily read it and understand what it tests.
+4. Each test is split into the standard arrange, act, and assert steps with helpful explanations.
+5. The tests cover complex functions likely affected by regressions.
+
+
+## Explanation
+Tests help us improve the quality of our software. However, tests carry maintenance costs, so **our tests should add value to our software**. In pursuit of this value:
+
+1. **Avoid overly-specific coverage goals.**  
+  E.g. 80% code coverage. Goals like this incentivize writing more tests than required to obtain confidence in the software.
+2. **Don't test what you don't own.**  
+  Testing languages, frameworks, and other third-party software indicates you don't trust the dependency you decided to rely on.
+3. **Avoid testing basic operations.**  
+  E.g. Testing an Eloquent model's relationships. Though you own the method on the model, in nearly every case, it merely consumes the framework.
+4. **Test the input and output, not the implementation.**  
+  Implemtation may change over time. Tests shouldn't fail after refactoring.
+
+Breaking these guidelines usually adds to maintenance costs without adding worthwhile value to our software.
+
+### Unit tests
+*Test individual pieces of code.*
+
+Unit tests verify a relatively small piece of code is doing what it's intended to do. They shouldn’t have dependencies on outside systems, because they test internal consistency, not that units play nicely with outside systems.
+
+#### When (not) to use
+Write unit tests for non-trivial pieces of isolated code or to help you break integration tests into manageable steps.
+
+Avoid writing setup-heavy unit tests, because they're brittle. If you find yourself writing a unit test that requires complex mocking or requires significantly more code than the covered function, you should most likely use an integration test instead or rethink your implementation. Mocking tightly couples code to tests, which prevents proper refactoring (improving code without changing tests).
```

Comment:
It might be nice to add some heuristics about how to organize the tests - e.g. when it's better to have many small tests vs. less tests with multiple assertions (and possibly setup/action/assertion blocks) in them.

Reply:
I don't think I'm opposed to this, but I think that gets more complex and opens up more room for disagreement, so I'd prefer to see something like that added as an amendment.

The arrange, act, and assert stages are already covered in the explanation for the good test example, and I think that's sufficient.

Reply:
What about testing methods, such as 'isPrimeNumber' (or 'shouldApplyDiscout' etc), would you have the Arrange/Act/Assert blocks as well and would have just a single call to 'isPrimeNumber' in each separate test or is it acceptable to have just one test with multiple calls to 'isPrimeNumber' and no Arrange/Act/Assert comments?

I guess my current issue is that the way it's currently structured it looks like a well-written article with a set of recommendations which is called a 'standard', without mentioning a lot of alternative approaches or going deeply into the various testing techniques. The way I see it an inexperienced developer would blindly follow it to the point as the 'one true way' after reading it (same as coding from a tutorial), while a slightly more experienced developer would receive comments about not following the standard (even when it's perfectly appropriate).

Perhaps it'd have been a better idea to heavily reduce the scope and have it much more open-ended with a list of links to learning materials at the end..

Reply:
I personally prefer to **always** have arrange/act/assert steps, but I'm trying to avoid prescribing my preferences too strongly.

As for not going more deeply into testing techniques, that's in an effort to adhere to the ["We prefer fewer, simpler standards" tenet](https://github.com/HighlandSolutions/development-standards/tree/master/standards#we-prefer-fewer-simpler-standards). I also believe this is much more practical. To describe techniques more deeply would require more time than I believe it's worth and make it significantly more difficult to reach a consensus. It makes more sense to me to start from a higher level where we can agree well enough on things. If it turns out things are too general to be helpful enough, we can [amend](https://github.com/HighlandSolutions/development-standards/tree/master/standards#amendment) later.

As for experienced vs. inexperienced developers, Highland currently doesn't really hire inexperienced developers, and even if we did, they'd be on a team with more experienced developers who would provide guidance. Any confusion seen there could lead to clarifying amendments.

If we just had links pointing elsewhere, that wouldn't be in our voice or control. It would also most likely require a lot of time reading things not particularly relevant to the work we do and contain conflicting info.

Basically, my goal/approach is to find the right balance of generality vs specificity to be helpful w/o losing much time to diminishing returns.